File: src/auth/reauthStore.ts
/**
 * Re-auth State Store
 *
 * Long-lived Redis storage keyed by re-auth identity.
 * TTL = fastAuthTokenTtlSeconds (default 48h).
 *
 * Redis-only (no Postgres) — if lost, device falls back to full auth.
 */

import crypto from 'node:crypto';
import { redis } from '../db/redis.js';
import { config } from '../config/index.js';

export interface ReauthState {
  subscriberId: string;
  imsi: string;
  mk: string;      // base64
  kAut: string;     // base64
  kEncr: string;    // base64
  counter: number;
  identity: string; // the re-auth ID itself
}

function reauthKey(reauthId: string): string {
  return `reauth:${reauthId}`;
}

/**
 * Generate a cryptographically random re-auth identity.
 */
export function generateReauthId(): string {
  return crypto.randomBytes(24).toString('base64url');
}

/**
 * Store re-auth state in Redis with TTL.
 */
export async function storeReauthState(state: ReauthState): Promise<void> {
  const key = reauthKey(state.identity);
  const fields: Record<string, string> = {
    subscriberId: state.subscriberId,
    imsi: state.imsi,
    mk: state.mk,
    kAut: state.kAut,
    kEncr: state.kEncr,
    counter: String(state.counter),
    identity: state.identity,
  };
  await redis.hset(key, fields);
  await redis.expire(key, config.fastAuthTokenTtlSeconds);
}

/**
 * Retrieve re-auth state by re-auth identity.
 * Returns null if not found or expired.
 */
export async function getReauthState(reauthId: string): Promise<ReauthState | null> {
  const key = reauthKey(reauthId);
  const data = await redis.hgetall(key);

  if (!data || Object.keys(data).length === 0) {
    return null;
  }

  return {
    subscriberId: data.subscriberId!,
    imsi: data.imsi!,
    mk: data.mk!,
    kAut: data.kAut!,
    kEncr: data.kEncr!,
    counter: parseInt(data.counter!, 10),
    identity: data.identity!,
  };
}

/**
 * Delete re-auth state (after rotation or expiry).
 */
export async function deleteReauthState(reauthId: string): Promise<void> {
  await redis.del(reauthKey(reauthId));
}

File: internal/eapaka/reauth_store.go
package eapaka

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"strconv"
	"time"

	"github.com/redis/go-redis/v9"
)

// ReauthState is the long-lived re-auth state stored in Redis.
type ReauthState struct {
	SubscriberID string
	IMSI         string
	MK           string // base64
	KAut         string // base64
	KEncr        string // base64
	Counter      int
	Identity     string // the re-auth ID itself
}

// ReauthStore manages long-lived re-auth state in Redis.
type ReauthStore struct {
	client *redis.Client
	ttl    time.Duration
}

// NewReauthStore creates a new ReauthStore.
func NewReauthStore(client *redis.Client, ttlSeconds int) *ReauthStore {
	return &ReauthStore{
		client: client,
		ttl:    time.Duration(ttlSeconds) * time.Second,
	}
}

func reauthStoreKey(reauthID string) string {
	return "reauth:" + reauthID
}

// GenerateReauthID generates a cryptographically random re-auth identity.
func GenerateReauthID() (string, error) {
	b := make([]byte, 24)
	if _, err := rand.Read(b); err != nil {
		return "", fmt.Errorf("crypto/rand: %w", err)
	}
	return base64.RawURLEncoding.EncodeToString(b), nil
}

// Store saves re-auth state in Redis with TTL.
func (s *ReauthStore) Store(ctx context.Context, state *ReauthState) error {
	key := reauthStoreKey(state.Identity)
	fields := map[string]interface{}{
		"subscriberId": state.SubscriberID,
		"imsi":         state.IMSI,
		"mk":           state.MK,
		"kAut":         state.KAut,
		"kEncr":        state.KEncr,
		"counter":      strconv.Itoa(state.Counter),
		"identity":     state.Identity,
	}

	pipe := s.client.Pipeline()
	pipe.HSet(ctx, key, fields)
	pipe.Expire(ctx, key, s.ttl)
	if _, err := pipe.Exec(ctx); err != nil {
		return fmt.Errorf("store reauth state: %w", err)
	}
	return nil
}

// Get retrieves re-auth state by identity. Returns nil if not found.
func (s *ReauthStore) Get(ctx context.Context, reauthID string) (*ReauthState, error) {
	key := reauthStoreKey(reauthID)
	data, err := s.client.HGetAll(ctx, key).Result()
	if err != nil {
		return nil, fmt.Errorf("get reauth state: %w", err)
	}
	if len(data) == 0 {
		return nil, nil
	}

	counter, _ := strconv.Atoi(data["counter"])

	return &ReauthState{
		SubscriberID: data["subscriberId"],
		IMSI:         data["imsi"],
		MK:           data["mk"],
		KAut:         data["kAut"],
		KEncr:        data["kEncr"],
		Counter:      counter,
		Identity:     data["identity"],
	}, nil
}

// Delete removes re-auth state.
func (s *ReauthStore) Delete(ctx context.Context, reauthID string) error {
	return s.client.Del(ctx, reauthStoreKey(reauthID)).Err()
}
